fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.lam = lam
+        self.mkmmd_loss_weight = mkmmd_loss_weight
+        self.feature_l2_norm = feature_l2_norm
+        self.beta_update_interval = beta_update_interval
+        self.mkmmd_loss = MkMmdLoss(device=self.device, minimize_type_two_error=True).to(self.device)
+
+        self.global_model: nn.Module
+        self.init_global_model: nn.Module
+
+    def get_optimizer(self, config: Config) -> Dict[str, Optimizer]:
+        """
+        Returns a dictionary with global and local optimizers with string keys 'global' and 'local' respectively.
+        """
+        raise NotImplementedError
+
+    def set_optimizer(self, config: Config) -> None:
+        """
+        Ditto requires an optimizer for the global model and one for the local model. This function simply ensures that
+        the optimizers setup by the user have the proper keys and that there are two optimizers.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        optimizers = self.get_optimizer(config)
+        assert isinstance(optimizers, dict) and set(("global", "local")) == set(optimizers.keys())
+        self.optimizers = optimizers
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        Then set initialized attribute to True.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        # Need to setup the global model here as well. It should be the same architecture as the local model so
+        # we reuse the get_model call. We explicitly send the model to the desired device. This is idempotent.
+        self.global_model = self.get_model(config).to(self.device)
+        # The rest of the setup is the same
+        super().setup_client(config)
+
+    def get_ditto_drift_loss(self) -> torch.Tensor:
+        """
+        Compute the L2 inner product between the initial global weights for the round and the current local model
+            weights. This loss function is added to the loss function for the local model when back propagating.
+
+        Returns:
+            torch.Tensor: Returns the L2 inner product between the initial global weights of the round and the current
+                local model weights.
+        """
+        assert self.init_global_model is not None and self.model is not None and self.lam is not None
+        # Using parameters to ensure the same ordering as exchange
+        local_model_weights = [layer_weights for layer_weights in self.model.parameters()]
+        # Detach the weights to ensure we don't compute gradients with respect to the tensors
+        initial_global_weights = [layer_weights.detach() for layer_weights in self.init_global_model.parameters()]
```

Comment:
Each time we form the loss (i.e. every batch), we're going to be recomputing this quantity, even though it never changes during a client-side training run. Perhaps we can do this one right after cloning and freezing the global model in `update_before_train`?